packages/firestore/src/local/indexeddb_persistence.ts

```diff
@@ -263,7 +282,18 @@ export class IndexedDbPersistence implements Persistence {
           currentPrimary.ownerId !== this.getZombiedClientId();
 
         if (currentLeaseIsValid) {
-          return this.isLocalClient(currentPrimary);
+          if (this.isLocalClient(currentPrimary)) {
+            return true;
+          }
+
+          if (!currentPrimary.allowTabSynchronization) {
```

Comment:
FWIW this seems like a non-obvious (and not 100% safe) place for this check.  It'd make more sense to me that we do this as part of runTransaction() so that every time we perform a transaction we make sure there's no external primary client that has allowTabSynchronization turned off.

As-is, a non-primary client could do up to 4 seconds of "corrupting" persistence before it notices that another tab is running without tab synchronization.

So it seems like we should either fix this or else consider it an intentional design choice and add some comments to explain.

Reply:
We need to do this both here and in `runTransaction`. I added a comment that should make the reasoning for this a little more obvious. In essence, we need to fail the `enablePersistence()` promise, and running this check in `canActAsPrimary` does this at startup while also blocking the AsyncQueue if this check were to ever fail at a later point.

As you pointed out, we do need to run this check for every transaction . I added this to `runTransaction` though as I do need it to ensure correctness.

I could also call the new method `verifyAllowTabSynchronization` at startup, but that is an additional read for every client, and will in practice not make a difference.